docs/jwst/user_documentation/running_pipeline_python.rst

```diff
@@ -529,3 +529,166 @@ individual step parameter must be set when using this method, or else the coded
 defaults will be used, which may be inappropriate for the dataset being processed.
 
 See :ref:`call_examples` for more information.
+
+
+.. _multiprocessing:
+
+Multiprocessing
+===============
+
+Multiprocessing is supported to speed up certain computationally-intensive steps
+in the pipeline, including :ref:`jump_step <jump_step>`,
+:ref:`ramp_fitting_step <ramp_fitting_step>`, and
+:ref:`wfss_contam_step <wfss_contam_step>`. The examples below show how
+multiprocessing can be enabled for these steps, as well as how to set up
+multiprocessing to simultaneously run the entire pipeline on multiple observations.
+
+Python's multiprocessing module explicitly imports and executes a script's
+`__main__` with each and every worker. If `__main__` is not present the behavior is
+undefined. Hence, Python will crash unless the multiprocess code in enclosed in a
+`__main__` block like this:
+
+
+::
+
+    import sys
+
+    def main():
+        [code used in multiprocessing]
+
+    if __name__ = '__main__':
+        sys.exit(main())
+
+
+There are a couple of scenarios to use multiprocessing with the pipeline:
+
+1.  Multiprocessing within a pipeline step. At the moment, the steps that
+support this are the :ref:`jump_step <jump_step>`,
+:ref:`ramp_fitting_step <ramp_fitting_step>`,
+and :ref:`wfss_contam_step <wfss_contam_step>`. To enable multiprocessing the
+optional parameters are `max_cores` for the ``jump`` step, and `maximum_cores`
+for the ``ramp_fitting`` and ``wfss_contam`` steps. These parameters can be
+set to `quarter`, `half`, `all`, or `none`, which is the default value.
+
+The following example turns on a step's multiprocessing option. Notice only
+one of the steps has multiprocessing turned on. We do not recommend to
+simultaneously enable both steps to do multiprocessing, as this may likely
+lead to running out of system memory.
+
+
+
+::
+
+    # SampleScript1
+
+    import os, sys
+    from jwst.pipeline import Detector1Pipeline
+
+    uncal_file = 'jw0000_0000_uncal.fits'
+    output_dir = '/my_project'
+
+    def main():
+        det1 = Detector1Pipeline()
+        parameter_dict = {"ramp_fit": {"maximum_cores": 'all'}}
+        det1.call(uncal_file, save_results=True, steps=parameter_dict, output_dir=output_dir)
+
+    if __name__ = '__main__':
+        sys.exit(main())
+
+
+2. Calling the pipeline using multiprocessing. The following example uses this
+option setting up a log file for each run of the pipeline and a text file with
+the full traceback in case there is a crash. Notice that the ``import`` statement
+of the pipeline is within the multiprocessing block that gets called by every
+worker. This is to avoid a known memory leak.
```

Comment:
Is the memory leak addressed by?
https://github.com/spacetelescope/jwst/pull/8343

Reply:
nope, memory leak still happening. Will add a couple of PRs to address this.

Reply:
Thanks for the response. Is there an example you can share that leaks memory? I tried the example in: https://github.com/spacetelescope/jwst/issues/8404 and wasn't able to replicate it locally. If it's similar enough to the linked issue feel free to follow up there.

Reply:
After a chat with Ned and Brett, we concluded that the semaphore warnings is not an issue at this moment, so I closed the PRs I had open. I think the issue is not necessarily related to https://github.com/spacetelescope/jwst/issues/8404, as it is resolved if the user encloses their code in a __main__ block. The semaphore warnings could be related to the usage to lots of memory by repeatedly opening datamodels (which creates a copy of the previous model). Brett is addressing this.